scripts/agd-builder.sh

```diff
@@ -0,0 +1,245 @@
+        }
+        ;;
+    esac
+    # Ensure minimum patch versions of Go environment
+    cd "$GOLANG_DIR"
+    if goversion=$(go version 2> /dev/null); then
+      goregexp='go version go([0-9]+)(.([0-9]+)(.([0-9]+))?)? '
+      [[ "$goversion" =~ $goregexp ]] || fatal "illegible go version '$goversion'"
+      golang_version_check "${BASH_REMATCH[1]}" "${BASH_REMATCH[3]}" "${BASH_REMATCH[5]}"
+      make compile-go
+    fi
+  )
+
+  if $need_nodejs; then
+    lazy_yarn() {
+      yarn --version 2> /dev/null 1>&2 || {
+        # Auto-download the Yarn version we need, if allowed.
+        $SKIP_DOWNLOAD || {
+          npm install -g yarn
+        }
+      }
+      yarn "$@"
+    }
+
+    # Check if any package.json is newer than the installation stamp.
+    # If some package.json is newer than the last yarn install, run yarn install
+    # UNTIL https://github.com/Agoric/agoric-sdk/issues/9209
+    stamp=$STAMPS/yarn-installed
+    if test -e "$stamp"; then
+      print=(-newer "$stamp")
+    else
+      print=()
+    fi
+    print+=(-print)
+
+    # Find the current list of package.jsons.
+    files=(package.json)
+    while IFS= read -r line; do
+      files+=("$line")
+    done < <(lazy_yarn -s workspaces info \
+      | sed -ne '/"location":/{ s/.*": "//; s!",.*!/package.json!; p; }')
+
+    src=$(find "${files[@]}" "${print[@]}" | head -1 || true)
+    test -z "$src" || {
+      echo "At least $src is newer than node_modules"
+      rm -f "$STAMPS/yarn-built"
+      lazy_yarn install
+    }
+
+    stamp=$STAMPS/yarn-built
+    if test ! -e "$stamp"; then
+      echo "Yarn packages need to be built"
+      lazy_yarn build
+    fi
+
+    # Simple test to see whether to build the GYP bindings.
+    # Check if any GYP source is newer than the destination.
+    stamp=$GOLANG_DIR/build/Release/agcosmosdaemon.node
+    if test -e "$stamp"; then
+      print=(-newer "$stamp")
+    else
+      print=()
+    fi
+    print+=(-print)
+
+    files=("$GOLANG_DIR/build/"*agcosmosdaemon.h)
+    src=$(find "${files[@]}" "${print[@]}" | head -1 || true)
+    test -z "$src" || {
+      echo "At least $src is newer than gyp bindings"
+      (cd "$GOLANG_DIR" && lazy_yarn build:gyp)
+    }
+  fi
+)
+
+# the xsnap binary lives in a platform-specific directory
+unameOut="$(uname -s)"
+case "${unameOut}" in
+  Linux*) platform=lin ;;
+  Darwin*) platform=mac ;;
+  *) platform=win ;;
+esac
+
+# check the xsnap version against our baked-in notion of what version we should be using
+xsnap_version=$("${thisdir}/../packages/xsnap/xsnap-native/xsnap/build/bin/${platform}/release/xsnap-worker" -n)
+[[ "${xsnap_version}" == "${XSNAP_VERSION}" ]] || fatal "xsnap version mismatch; expected ${XSNAP_VERSION}, got ${xsnap_version}"
+
+if $only_build; then
+  echo "Build complete." 1>&2
+  exit 0
+fi
+
+# Run the built Cosmos daemon.
+# shellcheck disable=SC2031
+export PATH="$thisdir/../packages/cosmic-swingset/bin:$PATH"
+exec "$thisdir/../$GOLANG_DAEMON" ${1+"$@"}
```

Comment:
consider moving this to `bin/agd` so this builder script's only job is to build

Reply:
Oh, good idea!